File: eau03.rb
# fibonacci
def log_and_die(error)
  puts error
  exit(1)
end

VALID_CHAR = %w[0 1 2 3 4 5 6 7 8 9].freeze
def digit?(str)
  return false if str.empty?

  start = if str[0] == '-'
            1
          elsif str[0] == '+'
            1
          else
            0
          end
  limit = str.length - 1
  (start..limit).each do |index|
    return false unless VALID_CHAR.include? str[index]
  end
  true
end

def fibonacci(number)
  return -1 if number.negative?

  accumulator = [0, 1]
  i = 0
  while i < number
    temp = accumulator[0]
    accumulator[0] = accumulator[1]
    accumulator[1] = accumulator[1] + temp
    i += 1
  end

  accumulator[0]
end

log_and_die 'Usage: eau03.rb <number>' if ARGV.length != 1
arg = ARGV[0]
log_and_die "#{arg} is not a valid number" unless digit?(arg)

number = Integer(arg)

puts fibonacci(number)

File: eau03.exs
# Fibonacci
defmodule Exercice do
  @spec validate_args(list(String.t())) ::
          {:ok, integer()}
          | {:error, :bad_args}
          | {:error, {:not_number, String.t()}}

  def validate_args([arg]) do
    case Integer.parse(arg) do
      {number, _rest} ->
        {:ok, number}

      :error ->
        {:error, {:not_number, arg}}
    end
  end

  def validate_args(_list), do: {:error, :bad_args}

  def fib(n) when n < 0, do: -1
  def fib(0), do: 0
  def fib(1), do: 1

  def fib(n) do
    fib(n - 1) + fib(n - 2)
  end
end

System.argv()
|> Exercice.validate_args()
|> case do
  {:ok, number} ->
    number
    |> Exercice.fib()
    |> IO.puts()

  {:error, :bad_args} ->
    IO.puts("Usage: elixir eau03.exs <number>")

  {:error, {:not_number, arg}} ->
    IO.puts("\"#{arg}\" is not a valid number")
end
